lib/galaxy/files/sources/__init__.py

```diff
@@ -296,7 +299,7 @@ def get_uri_root(self) -> str:
         """Return a prefix for the root (e.g. gxfiles://prefix/)."""
 
     @abc.abstractmethod
-    def list(
+    async def list(
```

Comment:
Please don't change this for existing interfaces. You can add a `<operation>_async` method where needed. We can't do blocking IO in async calls, but every plugin apart from yours will make blocking calls.

Reply:
**TL;DR**: _no public instances, hard to define abuse and to be abused, cannot be engineered much better without eLabFTW API changes_

eLabFTW's scope is one instance per lab or at most one per research institution. Unlike Galaxy, there are no public main instances. Either you host your own, or you use their [hosted offer](https://www.deltablot.com/elabftw/), which I assume is a single instance just for you (it costs 4985 € / yr) and is supposed to handle _"~128 active users max"_.

Therefore, there is no authority defining abuse that it makes sense to ask or that would ban your ip per se. It depends on how many resources the endpoint you have configured has at their disposal, what their admins consider reasonable, and whether they enforce rate limits or not.

Nevertheless, and foreseeing this problem, I included a hard limit [`MAX_CONCURRENT_REQUESTS: int = 75`](https://github.com/galaxyproject/galaxy/pull/19319/files#diff-226b24a98914f99eb1b630e5ddb06da305bb99bc19386adcd5e538cb1c595be3R88), setting the number to something that would not cause issues (for example timed out requests according to the [timeouts I also have defined](https://github.com/galaxyproject/galaxy/pull/19319/files#diff-226b24a98914f99eb1b630e5ddb06da305bb99bc19386adcd5e538cb1c595be3R90-R92)) on their [live demo site](https://demo.elabftw.net). To decrease the number of requests, I also avoid making requests if it is known that the experiment or resource has no attachments [`wrapped_entity.source["has_attachment"]`](https://github.com/galaxyproject/galaxy/pull/19319/files#diff-226b24a98914f99eb1b630e5ddb06da305bb99bc19386adcd5e538cb1c595be3R459).

From a software engineering standpoint, given the fact that the file sources API requires the plugin to be able to list everything on the server (`recursive=True`) and the pagination is client-side I guess I cannot do much better besides a global queue with a limit that applies to the sum of requests per endpoint from all users at any time.

From this point on, achieving a better result if the plugin is supposed to work in the case `recursive=True` becomes a _social engineering_ (from the point of view of our project) problem rather than a software engineering one, as [we'd need changes to the eLabFTW API](https://github.com/galaxyproject/galaxy/pull/19319/files#diff-226b24a98914f99eb1b630e5ddb06da305bb99bc19386adcd5e538cb1c595be3R18-R24). At best and enpoint [like this one](https://doc.elabftw.net/api/v2/#/User%20Uploads/read-user-uploads) but that lists all attachments _a user can access_, ideally takes offset, limit, sorting and the filtering parameters we need.

Reply:
That is why I think just not supporting `recursive=True` in those file sources that don't make sense would be the safest option. We could add a "support" flag, similar to what we already have for pagination, etc. here:

https://github.com/galaxyproject/galaxy/blob/b1217afc0cd42a280dfadcf6898d50b28c352e25/lib/galaxy/files/sources/__init__.py#L84

Then we can disable the recursive option in the UI and the API according to this flag.